### PR TITLE
Improving handling of Oauth exceptions

### PR DESCRIPTION
**Pull request by KurtStam**
Best effort parsing of the raw error message, with a fallback of sending the raw message. This message is set in the userMsgDetail.

https://github.com/syndesisio/syndesis-rest/issues/679
https://github.com/syndesisio/syndesis-rest/issues/652

Let's simply add more parsers as they different raw formats are found.

### CODE REVIEW

File: credential/src/test/java/io/syndesis/credential/CredentialsTest.java [hunk 1]
```diff
@@ -264,4 +267,29 @@ public void shouldFinishOAuth2Acquisition() {
         assertThat(finalFlowState)
             .isEqualTo(new OAuth2CredentialFlowState.Builder().createFrom(flowState).accessGrant(accessGrant).build());
     }
+    
+    @Test()
+    public void checkTwitterOAuthException() {
+        String consumerKey="NMqaca1bzXsOcZhP2XlwA";
+        String consumerSecret="VxNQiRLwwKVD0K9mmfxlTTbVdgRpriORypnUbHhxeQw";
```
Comment (zregvart) [inline credential/src/test/java/io/syndesis/credential/CredentialsTest.java]: are these the real deal?

File: credential/src/test/java/io/syndesis/credential/CredentialsTest.java [hunk 2]
```diff
@@ -264,4 +267,29 @@ public void shouldFinishOAuth2Acquisition() {
...
+        final OAuth1ConnectionFactory<?> oauth1 = new TwitterConnectionFactory(consumerKey, consumerSecret);
+        
+        String accessToken="26693234-W0YjxL9cMJrC0VZZ4xdgFMymxIQ10LeL1K8YlbBY";
+        String accessTokenSecret="BZD51BgzbOdFstWZYsqB5p5dbuuDV12vrOdatzhY4E";
```
Comment (zregvart) [inline credential/src/test/java/io/syndesis/credential/CredentialsTest.java]: ditto?

File: credential/src/test/java/io/syndesis/credential/CredentialsTest.java [hunk 3]
```diff
@@ -264,4 +267,29 @@ public void shouldFinishOAuth2Acquisition() {
...
+        when(locator.providerWithId("providerId"))
+            .thenReturn(new OAuth1CredentialProvider<>("providerId", oauth1, applicator));
+        try {
+            @SuppressWarnings("unused")
+            final AcquisitionFlow acquisition = credentials.acquire("providerId", URI.create("https://syndesis.io/api/v1/"),
+                URI.create("https://syndesis.io/ui#state"));
+            Assert.fail();
+        } catch (HttpClientErrorException exception) {
+            //The HttpClientErrorExceptionMapper will pull out the message from here.
+            Assert.assertTrue(exception.getResponseBodyAsString().contains("<error>"));
```
Comment (zregvart) [inline credential/src/test/java/io/syndesis/credential/CredentialsTest.java]: Perhaps use AssertJ for asserting exception is thrown?

https://joel-costigliola.github.io/assertj/assertj-core-features-highlight.html#exception-assertion

File: rest/src/main/java/io/syndesis/rest/v1/handler/exception/HttpClientErrorExceptionMapper.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.HttpClientErrorException;
+
+@Component
+@Provider
+public class HttpClientErrorExceptionMapper implements ExceptionMapper<HttpClientErrorException> {
+
+    @Override
+    public Response toResponse(HttpClientErrorException exception) {
+        String message = exception.getMessage() +
```
Comment (zregvart) [inline rest/src/main/java/io/syndesis/rest/v1/handler/exception/HttpClientErrorExceptionMapper.java]: Perhaps we should check for the content type is XML?

https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/client/RestClientResponseException.html#getResponseHeaders--

File: credential/src/test/java/io/syndesis/credential/CredentialsTest.java (hunk 1 above)
Comment (KurtStam) [reply credential/src/test/java/io/syndesis/credential/CredentialsTest.java]: Not quite sure. I used the ones I found here: https://github.com/syndesisio/syndesis-rest/blob/master/runtime/src/test/resources/valid-twitter-keys.properties. Now that I captured an error we probably don't even need them anymore.

File: rest/src/main/java/io/syndesis/rest/v1/handler/exception/HttpClientErrorExceptionMapper.java (hunk 4 above)
Comment (KurtStam) [reply rest/src/main/java/io/syndesis/rest/v1/handler/exception/HttpClientErrorExceptionMapper.java]: y - I'd like to see what SF does

File: credential/src/test/java/io/syndesis/credential/CredentialsTest.java (hunk 1 above)
Comment (zregvart) [reply credential/src/test/java/io/syndesis/credential/CredentialsTest.java]: Cool ;)

File: rest/src/main/java/io/syndesis/rest/v1/handler/exception/HttpClientErrorExceptionMapper.java (hunk 4 above)
Comment (zregvart) [reply rest/src/main/java/io/syndesis/rest/v1/handler/exception/HttpClientErrorExceptionMapper.java]: For Salesforce we don't have that call, it's OAuth1 only and Salesforce is OAuth2 and there is a call at the end when we get an OK response from the OAuth callback already.

**Comment by KurtStam** [comment]
retest please

**Comment by KurtStam** [comment]
retest this please

**Comment by KurtStam** [comment]
retest this please

**Comment by zregvart** [comment]
Hey @KurtStam I PR'd your PR see https://github.com/KurtStam/syndesis-rest/pull/1 😄 seemed easier that leaving a bunch of review comments here

**Comment by rhuss** [comment]
Good we get the PR in asap ? I plan to go TP2 maybe already today ...

**Comment by zregvart** [comment]
> Good we get the PR in asap ? I plan to go TP2 maybe already today ...

Let's give @KurtStam a chance to review my comments via code PR, as is it is good to go and we could address those bits in another PR after if we're in a hurry.

**Comment by KurtStam** [comment]
@zregvart I reviewed and merged. It's in my branch but this PR is not updating..

**Comment by KurtStam** [comment]
retest this please

**Comment by chirino** [comment]
Manual merging since this is just waiting for Jenkins CI and that seems a little broken at the moment.

**Pull request merged**